### PR TITLE
FIX: be more careful about coercing unit-full containers to ndarray

### LINKED ISSUE #21669: [Bug]: Matplotlib 3.5 breaks unyt integration of error bars
### Bug summary

Discovered this in unyt's daily CI jobs: https://github.com/yt-project/unyt/actions

### Code for reproduction

```python
import matplotlib.pyplot as plt
import unyt as un

un.matplotlib_support.enable()
fig, ax = plt.subplots()

x = un.unyt_array([1, 2], "cm")
y = un.unyt_array([3, 4], "kg")
yerr = un.unyt_array([0.1, 0.2], "kg")
ax.errorbar(x, y, yerr=(yerr, yerr))
```

### Actual outcome

```
UnitOperationError: The <ufunc 'subtract'> operator for unyt_arrays with units "kg" 
(dimensions "(mass)") and "dimensionless" (dimensions "1") is not well defined.
```

<details> <summary>detailed raceback</summary>

```
╭──────────────────────────── Traceback (most recent call last) ────────────────────────────╮
│                                                                                           │
│ /private/tmp/matplotlib/t.py:10 in <module>                                               │
│                                                                                           │
│    7 x = un.unyt_array([1, 2], "cm")                                                      │
│    8 y = un.unyt_array([3, 4], "kg")                                                      │
│    9 yerr = un.unyt_array([0.1, 0.2], "kg")                                               │
│ ❱ 10 ax.errorbar(x, y, yerr=yerr)                                                         │
│   11                                                                                      │
│ /private/tmp/matplotlib/lib/matplotlib/__init__.py:1348 in inner                          │
│                                                                                           │
│   1345 │   @functools.wraps(func)                                                         │
│   1346 │   def inner(ax, *args, data=None, **kwargs):                                     │
│   1347 │   │   if data is None:                                                           │
│ ❱ 1348 │   │   │   return func(ax, *map(sanitize_sequence, args), **kwargs)               │
│   1349 │   │                                                                              │
│   1350 │   │   bound = new_sig.bind(ax, *args, **kwargs)                                  │
│   1351 │   │   auto_label = (bound.arguments.get(label_namer)                             │
│                                                                                           │
│ /private/tmp/matplotlib/lib/matplotlib/axes/_axes.py:3474 in errorbar                     │
│                                                                                           │
│   3471 │   │   │   │   │   │   xo, yo, marker='|', **eb_cap_style))                       │
│   3472 │   │                                                                              │
│   3473 │   │   if yerr is not None:                                                       │
│ ❱ 3474 │   │   │   lower, upper = extract_err('y', yerr, y, lolims, uplims)               │
│   3475 │   │   │   barcols.append(self.vlines(                                            │
│   3476 │   │   │   │   *apply_mask([x, lower, upper], everymask), **eb_lines_style))      │
│   3477 │   │   │   # select points without upper/lower limits in y and                    │
│                                                                                           │
│ /private/tmp/matplotlib/lib/matplotlib/axes/_axes.py:3434 in extract_err                  │
│                                                                                           │
│   3431 │   │   │   │   │   f"'{name}err' (shape: {np.shape(err)}) must be a scalar "      │
│   3432 │   │   │   │   │   f"or a 1D or (2, n) array-like whose shape matches "           │
│   3433 │   │   │   │   │   f"'{name}' (shape: {np.shape(data)})") from None               │
│ ❱ 3434 │   │   │   return data - low * ~lolims, data + high * ~uplims  # low, high        │
│   3435 │   │                                                                              │
│   3436 │   │   if xerr is not None:                                                       │
│   3437 │   │   │   left, right = extract_err('x', xerr, x, xlolims, xuplims)              │
│                                                                                           │
│ /Users/robcleme/.pyenv/versions/3.9.7/envs/tmp_unit/lib/python3.9/site-packages/unyt/arra │
│ y.py:1764 in __array_ufunc__                                                              │
│                                                                                           │
│   1761 │   │   │   │   │   │   │   │   else:                                              │
│   1762 │   │   │   │   │   │   │   │   │   raise UnitOperationError(ufunc, u0, u1)        │
│   1763 │   │   │   │   │   │   else:                                                      │
│ ❱ 1764 │   │   │   │   │   │   │   raise UnitOperationError(ufunc, u0, u1)                │
│   1765 │   │   │   │   │   conv, offset = u1.get_conversion_factor(u0, inp1.dtype)        │
│   1766 │   │   │   │   │   new_dtype = np.dtype("f" + str(inp1.dtype.itemsize))           │
│   1767 │   │   │   │   │   conv = new_dtype.type(conv)                                    │
╰───────────────────────────────────────────────────────────────────────────────────────────╯
UnitOperationError: The <ufunc 'subtract'> operator for unyt_arrays with units "kg" 
(dimensions "(mass)") and "dimensionless" (dimensions "1") is not well defined.
```

</details>

### Expected outcome

This is the output figure with matplotlib 3.4.3
unyt_err

### Additional information

The regression bisects to #19526 (https://github.com/matplotlib/matplotlib/pull/19526/commits/8cd22b4d5840b08cd42f793f5a30e23595aa44f2)

I note that if replace the last line with
```
ax.errorbar(x, y, yerr=yerr)
```
(passing a single array instead of a tuple), it doesn't break on the main branch.

### Operating system

Unbuntu

### Matplotlib Version

3.5.0

### Matplotlib Backend

not relevant (I get an identical error with the 'agg' backend)

### Python version

tested on 3.7 to 3.9

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

closes #21669

Tested manually locally via:

```python
import matplotlib.pyplot as plt
import unyt as un

un.matplotlib_support.enable()
fig, ax = plt.subplots()

x = un.unyt_array([1, 2], "cm")
y = un.unyt_array([3, 4], "kg")
yerr = un.unyt_array([0.1, 0.2], "kg")
ax.errorbar(x, y, yerr=(yerr, yerr))

plt.show()

```

We do not currently have _any_ tests of errorbar + units, in interest of getting 3.5.1 out with this fix and the colorbar fixes I move we merge this without a test and kick this particular technical debt can down the road as making sure we have good test coverage is in-scope for the NASA grant.

## PR Checklist

**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -3282,10 +3282,34 @@ def errorbar(self, x, y, yerr=None, xerr=None,
             x = np.asarray(x, dtype=object)
         if not isinstance(y, np.ndarray):
             y = np.asarray(y, dtype=object)
+
+        def _upcast_errors(err):
```
Comment (jklymak) [inline lib/matplotlib/axes/_axes.py:3286]: Maybe just `upcast_err`?  Otherwise it sounds like you are going to deal with capital-E Errors...

Comment (tacaswell) [reply lib/matplotlib/axes/_axes.py]: done.

**Review by jklymak — DISMISSED** [review]

**Review by neutrinoceros — APPROVED** [review]
I confirm that this patch
- [x] fixes the addressed issue
- [x] fixes unyt's failing test
- [x] doesn't cause new test failures in unyt

thank you very much for this patch !

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -3282,10 +3282,56 @@ def errorbar(self, x, y, yerr=None, xerr=None,
...
+                    len(err) > 0 and
+                    # and the first element is an array sub-class use
+                    # safe_first_element because getitem is index-first not
+                    # location first on pandas objects so err[0] almost always
+                    # fails.
+                    isinstance(cbook.safe_first_element(err), np.ndarray)
+            ):
+                # grab the type of the first element, we will try to promote
+                # the outer container to match the inner container
+                atype = type(cbook.safe_first_element(err))
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py:3323]: This would be a fun place to use the walrus operator, but we do not have access to that yet 😞

**Comment by tacaswell** [comment]
@efiring @dopplershift can you review this (pinging you based on timezone ;) ) so @QuLogic can get 3.5.1 out tonight?

I would also leave this to @QuLogic 's discretion to merge on one (his) review.

**Review by dopplershift — APPROVED** [review]
This seems reasonable. I *think* it would work for Pint, but honestly I haven't had need for errorbars, so I've never tested.

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -3282,10 +3282,56 @@ def errorbar(self, x, y, yerr=None, xerr=None,
...
         if not isinstance(y, np.ndarray):
             y = np.asarray(y, dtype=object)
+
+        def _upcast_err(err):
+            """
+            Safely handle tuple of containers that carry units.
+
+            If the units are carried on the values then casting to object
+            arrays preserves the units, but if the units are on the containers
+            this will not work.
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py:3292]: I find the docstring confusing, mixing the what and the why in a way that I can't sort out.
I suggest simply stating what it does under each case it handles.

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -3282,10 +3282,56 @@ def errorbar(self, x, y, yerr=None, xerr=None,
...
+            We defer coercing the units to be consistent to the underlying unit
+            library (and implicitly the broadcasting).
+
+            If we do not have a tuple of nested numpy array (subclasses),
+            fallback to casting to an object array.
+
+            """
+
+            # we are here because we the container is not a numpy array, but it
+            # _is_ iterable (likely a list or a tuple but maybe something more
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py:3307]: But at this point you don't know it is iterable--you check it.  If it is not iterable, you still wrap it in an object array.

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -3282,10 +3282,56 @@ def errorbar(self, x, y, yerr=None, xerr=None,
...
+                    # fails.
+                    isinstance(cbook.safe_first_element(err), np.ndarray)
+            ):
+                # grab the type of the first element, we will try to promote
+                # the outer container to match the inner container
+                atype = type(cbook.safe_first_element(err))
+                # you can not directly pass data to the init of `np.ndarray`
+                if atype is np.ndarray:
+                    return np.asarray(err, dtype=object)
+                # but you can for unyt and astropy uints
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py:3327]: astropy "units", I presume

**Review by efiring — APPROVED** [review]
Suggested clarification of docstring and comments need not delay merging.

**Pull request merged**